src/Symfony/Component/Form/Extension/Csrf/EventListener/CsrfValidationListener.php

```diff
@@ -59,14 +60,15 @@ public function preSubmit(FormEvent $event)
         if ($form->isRoot() && $form->getConfig()->getOption('compound') && !$postRequestSizeExceeded) {
             $data = $event->getData();
 
-            if (!isset($data[$this->fieldName]) || !$this->tokenManager->isTokenValid(new CsrfToken($this->tokenId, $data[$this->fieldName]))) {
+            $csrfToken = new CsrfToken($this->tokenId, $data[$this->fieldName]);
+            if (!isset($data[$this->fieldName]) || !$this->tokenManager->isTokenValid($csrfToken)) {
                 $errorMessage = $this->errorMessage;
 
                 if (null !== $this->translator) {
                     $errorMessage = $this->translator->trans($errorMessage, array(), $this->translationDomain);
                 }
 
-                $form->addError(new FormError($errorMessage));
+                $form->addError(new FormError($errorMessage, null, array(), null, new ValidationError($csrfToken)));
```

Comment:
isnt `$csrfToken` the actual _cause_ here? what's the point of `getCause()->getCsrfToken() vs. getCause()` :thinking:

Reply:
I added a comment [here](https://github.com/symfony/symfony/pull/28564#discussion_r221456445). If I understand correctly it seems that CSRF token does not always exist. So I can not set CsrfToken object directly as cause.

Reply:
now it's declared at L63 :)